decomposer/neuro_san_solver.py

```diff
@@ -203,122 +327,6 @@ def _solve_atomic_with_voting(self, problem: str) -> tuple[str, list[str], list[
 
         return solutions[winner_idx], finals, votes, winner_idx, solutions
 
-    def solve_trace(self, problem: str, depth: int, max_depth: int, path: str) -> tuple[str, dict]:
-        """
-        Internal recursive solver that returns (response, trace_node).
-        Builds a complete trace tree of the decomposition process.
-        """
-        logging.info(f"[solve] depth={depth} path={path} problem: {problem[:120]}{'...' if len(problem) > 120 else ''}")
-
-        node = {
-            "depth": depth,
-            "path": path,
-            "problem": problem,
-            "decomposition": None,
-            "children": [],
-            "sub_finals": None,
-            "composition": None,
-            "response": None,
-            "final": None,
-            "final_num": None,
-            "error": None,
-        }
-
-        if depth >= max_depth:
-            logging.info(f"[solve] depth={depth} -> atomic (max depth)")
-            resp, finals, votes, winner_idx, solutions = self._solve_atomic_with_voting(problem)
-            node["response"] = resp
-            node["final"] = finals[winner_idx]
-            node["atomic"] = {
-                "atomic_candidates": finals,
-                "atomic_votes": votes,
-                "atomic_winner_idx": winner_idx,
-                "final_choice": finals[winner_idx],
-            }
-            return resp, node
-
-        p1, p2, c, decomp_meta = self.decompose(problem)
-
-        if not p1 or not p2 or not c:
-            logging.info(f"[solve] depth={depth} -> atomic (no decomp)")
-            if decomp_meta:
-                node["decomposition"] = {**decomp_meta, "decision": "no_decomposition"}
-            resp, finals, votes, winner_idx, solutions = self._solve_atomic_with_voting(problem)
-            node["response"] = resp
-            node["final"] = finals[winner_idx]
-            node["atomic"] = {
-                "atomic_candidates": finals,
-                "atomic_votes": votes,
-                "atomic_winner_idx": winner_idx,
-                "final_choice": finals[winner_idx],
-            }
-            return resp, node
-
-        logging.info(f"[solve] depth={depth} using decomposition")
-        node["decomposition"] = decomp_meta
-
-        s1_resp, s1_node = self.solve_trace(p1, depth + 1, max_depth, f"{path}.0")
-        s2_resp, s2_node = self.solve_trace(p2, depth + 1, max_depth, f"{path}.1")
-        node["children"] = [s1_node, s2_node]
-
-        s1 = self.extract_final(s1_resp)
-        s2 = self.extract_final(s2_resp)
-        node["sub_finals"] = {"s1_final": s1, "s2_final": s2}
```

Comment:
Note the calls to extract_final() from the response values from solve_trace in the original.